### PR TITLE
feat: AST-based ORDER BY stripping via new @memberjunction/sql-parser package

### LINKED ISSUE #2175: Replace regex-based ORDER BY stripping with AST-based approach using node-sql-parser
## Problem

The `stripTrailingOrderBy` method in `QueryCompositionEngine` ([queryCompositionEngine.ts:700-729](packages/MJCore/src/generic/queryCompositionEngine.ts#L700-L729)) uses regex to remove trailing `ORDER BY` clauses from SQL being wrapped in CTEs. This is fragile — `ORDER BY` can appear in:

- **Window functions**: `ROW_NUMBER() OVER (ORDER BY ...)`
- **Aggregate functions**: `STRING_AGG(...) WITHIN GROUP (ORDER BY ...)`
- **Subqueries**: `WHERE x IN (SELECT ... ORDER BY ...)`
- **String literals**: `'...ORDER BY...'`
- **SQL comments**: `-- ORDER BY ...`

A regex matching `ORDER BY ... $` risks mangling queries with these patterns. SQL Server disallows `ORDER BY` in CTEs unless `TOP`, `OFFSET`, or `FOR XML` is present, so the stripping must be precise.

### Current regex implementation

```typescript
private stripTrailingOrderBy(sql: string): string {
    const trimmed = sql.trimEnd();
    // Legal ORDER BY — keep it
    if (/\bTOP\s+\d/i.test(trimmed) || /\bOFFSET\s+\d/i.test(trimmed) || /\bFOR\s+XML\b/i.test(trimmed)) {
        return sql;
    }
    const orderByMatch = trimmed.match(/\bORDER\s+BY\s+[\s\S]+$/i);
    if (!orderByMatch) return sql;
    // Paren-depth check (basic protection for subqueries)
    const beforeOrderBy = trimmed.substring(0, orderByMatch.index);
    let parenDepth = 0;
    for (const ch of beforeOrderBy) {
        if (ch === '(') parenDepth++;
        else if (ch === ')') parenDepth--;
    }
    if (parenDepth !== 0) return sql;
    return trimmed.substring(0, orderByMatch.index).trimEnd();
}
```

The paren-depth check helps but doesn't cover window functions at the end of SELECT lists, WITHIN GROUP clauses, or string literals.

## Proposed Solution

Replace regex with **AST-based parsing** using `node-sql-parser`, which is already a dependency in the codebase ([MJCoreEntitiesServer/src/custom/sql-parser.ts](packages/MJCoreEntitiesServer/src/custom/sql-parser.ts)).

### Strategy (4-tier)

1. **Fast exit**: No `ORDER BY` keyword in SQL → return as-is
2. **Legal check**: `TOP` / `OFFSET` / `FOR XML` / `LIMIT` present → ORDER BY is legal, return as-is
3. **AST path** (happy path): Parse with `node-sql-parser`, null out `ast.orderby`, reconstruct with `sqlify()`. **This should be as robust as possible — handle Nunjucks template syntax by pre-processing templates before parsing, and cover edge cases thoroughly to minimize fallback usage.**
4. **Fallback** (AST fails on truly exotic SQL): Inject `OFFSET 0 ROWS` (SQL Server) — semantically neutral, makes ORDER BY legal in CTEs without changing results. For PostgreSQL, ORDER BY in CTEs is legal, so no-op. **Note**: `OFFSET 0 ROWS` switches SQL Server into "paging mode" internally, which _could_ affect query plan shape on very large CTEs. This is a last resort and should include a code comment explaining why it exists.

### Architecture: New `@memberjunction/sql-parser` Package

**Why a new package?**
- **MJCore must stay lean** (4 deps: `@memberjunction/global`, `rxjs`, `zod`, `debug`). Adding `node-sql-parser` (~2MB) would bloat Angular bundles since MJCore is imported by browser apps.
- **`SQLParser` was already designed standalone** — zero MJ imports, just `node-sql-parser` + `nunjucks`.
- **Multiple packages need it**: MJCoreEntitiesServer (entity extraction), CodeGenLib (view parsing), now QueryCompositionEngine (ORDER BY stripping).

**Dependencies**: `node-sql-parser`, `nunjucks`, and `@memberjunction/core` (for `DatabasePlatform` type and `LogStatusEx`). MJCore is tiny and already in the transitive dependency tree of every consumer, so this adds zero real weight.

**Pluggable registration pattern:**
- MJCore defines a `CteOrderByStripperFn` type and `RegisterOrderByStripper()` static method on `QueryCompositionEngine`
- Default: existing regex implementation (no new deps for MJCore)
- Server-side: `GenericDatabaseProvider` module imports `stripOrderByForCTE` from `@memberjunction/sql-parser` and registers it at module load time (before any queries execute)
- Browser stays unaffected — regex fallback just works

### ⚠️ Breaking Change

Moving `SQLParser` out of `@memberjunction/core-entities-server` and into the new `@memberjunction/sql-parser` package is a **breaking change** for any external consumer importing `SQLParser` from `@memberjunction/core-entities-server`. This will be treated as a patch — consumers must update their import paths to `@memberjunction/sql-parser`.

### Step 1: Create `packages/SQLParser/` package

```
packages/SQLParser/
  package.json          # deps: node-sql-parser, nunjucks, @memberjunction/core
  tsconfig.json
  vitest.config.ts
  src/
    index.ts            # Public exports
    sql-parser.ts       # MOVED from MJCoreEntitiesServer (unchanged logic)
    cte-order-by-stripper.ts   # NEW: AST-based ORDER BY stripping
    __tests__/
      cte-order-by-stripper.test.ts
```

Dependencies: `node-sql-parser@^5.4.0`, `nunjucks@^3.2.4`, `@memberjunction/core` (for `DatabasePlatform`, `LogStatusEx`).

The `stripOrderByForCTE` function uses the `DatabasePlatform` type from MJCore directly (not duplicated string literals). Maps to parser dialects: `'TransactSQL'` / `'postgresql'`.

### Step 2: Make `QueryCompositionEngine` stripper pluggable

**File**: `packages/MJCore/src/generic/queryCompositionEngine.ts`

- Export new type: `CteOrderByStripperFn = (sql: string, platform: DatabasePlatform) => string`
- Add static `RegisterOrderByStripper(fn)` method
- Rename current regex impl to `regexStripTrailingOrderBy` (kept as default fallback)
- Update `stripTrailingOrderBy` to delegate to registered function if one exists
- Thread `platform` parameter through `assembleCTEs(cteEntries, mainSQL, platform)` — `platform` is already available from `ResolveComposition` ([providerBase.ts:1375](packages/MJCore/src/generic/providerBase.ts#L1375))

### Step 3: Register AST stripper in `GenericDatabaseProvider`

**File**: `packages/GenericDatabaseProvider/src/GenericDatabaseProvider.ts`

- Add `@memberjunction/sql-parser` to `package.json` dependencies
- Register the AST stripper at module level (top of file, after imports):
  ```typescript
  import { stripOrderByForCTE } from '@memberjunction/sql-parser';
  import { QueryCompositionEngine } from '@memberjunction/core';

  // Module-level registration — runs when this module is first imported,
  // before any queries execute. Replaces the regex fallback in MJCore
  // with the AST-based implementation from @memberjunction/sql-parser.
  QueryCompositionEngine.RegisterOrderByStripper(stripOrderByForCTE);
  ```
- Registration is idempotent, harmless to call multiple times

### Step 4: Migrate existing SQLParser consumers

**4a. MJCoreEntitiesServer:**
- Delete `src/custom/sql-parser.ts` (moved to new package)
- Remove `export * from './custom/sql-parser'` from `src/index.ts` (line 16)
- Add `@memberjunction/sql-parser` to `package.json`
- Update `MJQueryEntityServer.server.ts` import (line 23): `import { SQLParser } from "@memberjunction/sql-parser"`

**4b. CodeGenLib:**
- Update `src/Database/manage-metadata.ts` import (line 10): `import { SQLParser } from "@memberjunction/sql-parser"`
- Add `@memberjunction/sql-parser` to `package.json`

### Step 5: Unit tests

**New tests** in `packages/SQLParser/src/__tests__/cte-order-by-stripper.test.ts`:
- Basic ORDER BY removal (AST path)
- Multi-column ORDER BY with ASC/DESC
- Preservation: TOP, OFFSET, FOR XML, LIMIT
- Window function: `ROW_NUMBER() OVER (ORDER BY ...)` — outer stripped, window preserved
- `STRING_AGG ... WITHIN GROUP (ORDER BY ...)` — outer stripped, aggregate preserved
- Subquery ORDER BY preserved
- String literal containing "ORDER BY" — not affected
- Nunjucks templates — pre-processed before AST parsing (AST path should succeed, not fallback)
- Cross-platform: both `'sqlserver'` and `'postgresql'`
- Edge cases: empty string, no ORDER BY, trailing semicolons

**Updated tests** in `packages/MJCore/src/__tests__/queryCompositionEngine.test.ts`:
- Add "Pluggable ORDER BY stripper" describe block: verify registration, delegation, and fallback

### Step 6 (Optional): Reusable query ORDER BY validation

**File**: `packages/MJCoreEntitiesServer/src/custom/MJQueryEntityServer.server.ts`

Add a verbose-only log via `LogStatusEx` in `Save()` when a Reusable query has a trailing ORDER BY without TOP/OFFSET. Informational only — doesn't block saves, only appears when `MJ_VERBOSE` is enabled:

```typescript
import { LogStatusEx } from '@memberjunction/core';

LogStatusEx({
    message: `Query "${this.Name}" has a trailing ORDER BY without TOP/OFFSET — this will be stripped when used in CTE composition`,
    verboseOnly: true,
    category: 'QueryValidation'
});
```

## Files Modified/Created

| File | Action |
|------|--------|
| `packages/SQLParser/` (entire directory) | **CREATE** — new package |
| `packages/SQLParser/src/cte-order-by-stripper.ts` | **CREATE** — AST-based ORDER BY stripping |
| `packages/SQLParser/src/sql-parser.ts` | **MOVE** from MJCoreEntitiesServer |
| `packages/MJCore/src/generic/queryCompositionEngine.ts` | **MODIFY** — pluggable stripper |
| `packages/GenericDatabaseProvider/src/GenericDatabaseProvider.ts` | **MODIFY** — register AST stripper |
| `packages/GenericDatabaseProvider/package.json` | **MODIFY** — add sql-parser dep |
| `packages/MJCoreEntitiesServer/src/custom/sql-parser.ts` | **DELETE** — moved to new package |
| `packages/MJCoreEntitiesServer/src/index.ts` | **MODIFY** — remove sql-parser export |
| `packages/MJCoreEntitiesServer/src/custom/MJQueryEntityServer.server.ts` | **MODIFY** — update import |
| `packages/MJCoreEntitiesServer/package.json` | **MODIFY** — add sql-parser dep |
| `packages/CodeGenLib/src/Database/manage-metadata.ts` | **MODIFY** — update import |
| `packages/CodeGenLib/package.json` | **MODIFY** — add sql-parser dep |

## Verification

1. `cd packages/SQLParser && npm run build && npm run test`
2. `cd packages/MJCore && npm run build && npm run test`
3. `cd packages/GenericDatabaseProvider && npm run build && npm run test`
4. `cd packages/MJCoreEntitiesServer && npm run build`
5. `cd packages/CodeGenLib && npm run build`

## Existing Code References

- **Current regex stripper**: [queryCompositionEngine.ts:700-729](packages/MJCore/src/generic/queryCompositionEngine.ts#L700-L729)
- **CTE assembly (calls stripper)**: [queryCompositionEngine.ts:673-692](packages/MJCore/src/generic/queryCompositionEngine.ts#L673-L692)
- **ResolveComposition entry point**: [queryCompositionEngine.ts:187](packages/MJCore/src/generic/queryCompositionEngine.ts#L187)
- **Existing SQLParser class**: [MJCoreEntitiesServer/src/custom/sql-parser.ts](packages/MJCoreEntitiesServer/src/custom/sql-parser.ts) — 432 lines, zero MJ deps, uses `node-sql-parser` + `nunjucks`
- **SQLParser usage in MJQueryEntityServer**: [MJQueryEntityServer.server.ts:385](packages/MJCoreEntitiesServer/src/custom/MJQueryEntityServer.server.ts#L385) — `SQLParser.ParseWithTemplatePreprocessing(this.SQL)`
- **SQLParser usage in CodeGenLib**: [manage-metadata.ts:1144](packages/CodeGenLib/src/Database/manage-metadata.ts#L1144) — `SQLParser.Parse(viewDefinition)`
- **DatabasePlatform type**: [platformSQL.ts](packages/MJCore/src/generic/platformSQL.ts) — `'sqlserver' | 'postgresql'`
- **Composition engine instance**: [providerBase.ts:1344](packages/MJCore/src/generic/providerBase.ts#L1344) — `private _compositionEngine = new QueryCompositionEngine()`
- **Platform threading**: [providerBase.ts:1375](packages/MJCore/src/generic/providerBase.ts#L1375) — `this.PlatformKey` passed to `ResolveComposition`
- **GenericDatabaseProvider class**: [GenericDatabaseProvider.ts:105](packages/GenericDatabaseProvider/src/GenericDatabaseProvider.ts#L105)
- **processQueryParameters**: [GenericDatabaseProvider.ts:2176](packages/GenericDatabaseProvider/src/GenericDatabaseProvider.ts#L2176)

## Related

- PR #2173 — Added ORDER BY stripping regex and query composition features
- `node-sql-parser` v5.4.0 already in dependency tree via MJCoreEntitiesServer

### PR DESCRIPTION
**Pull request by CaelebB-BC**
## Summary

- **New `@memberjunction/sql-parser` package** — houses the `SQLParser` class (moved from MJCoreEntitiesServer) and a new `stripOrderByForCTE()` function that uses `node-sql-parser` AST to strip trailing `ORDER BY` from CTE bodies
- **Fixes fragile regex stripping** — the old regex `\bORDER\s+BY\s+[\s\S]+$` could mangle queries with window functions (`ROW_NUMBER() OVER (ORDER BY ...)`), `STRING_AGG WITHIN GROUP`, subqueries, and string literals
- **Adds `AllowsOrderByInCTE` to `SQLDialect`** — platform CTE rules checked via dialect abstraction instead of hardcoded string comparisons (SQL Server: `false`, PostgreSQL: `true`)

## Strategy (3-tier)

1. **Fast exit** — no `ORDER` keyword, or dialect allows ORDER BY in CTEs (PostgreSQL)
2. **AST path** — parse with `node-sql-parser`, check `top`/`limit`/`for` AST nodes for legal ORDER BY, null out `orderby` if not legal, regenerate via `sqlify()`
3. **Regex fallback** — paren-depth heuristic for SQL the parser can't handle (Nunjucks templates)

## Breaking Change

`SQLParser` is no longer exported from `@memberjunction/core-entities-server`. Consumers must update imports:

```typescript
// Before
import { SQLParser } from '@memberjunction/core-entities-server';
// After
import { SQLParser } from '@memberjunction/sql-parser';
```

## Test plan

- [x] 78 unit tests in `packages/SQLParser/` — window functions, TOP/OFFSET/FOR XML, subqueries, string literals, Nunjucks, UNION, JOINs, GROUP BY, PostgreSQL pass-through, real-world patterns
- [x] 210 integration tests in `packages/GenericDatabaseProvider/` — composition engine with AST stripping, window function preservation, PostgreSQL CTE pass-through
- [x] `packages/MJCoreEntitiesServer` and `packages/CodeGenLib` build successfully with updated imports

Closes #2175

### CODE REVIEW

File: packages/SQLParser/src/cte-order-by-stripper.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+ * Attempts to strip the top-level ORDER BY clause using AST parsing.
+ * Checks AST nodes for TOP, OFFSET/FETCH, and FOR XML to determine if ORDER BY is legal.
+ * Returns null if the parser cannot handle the SQL (caller should fall back to regex).
+ */
+function stripOrderByViaAST(sql: string): string | null {
+    try {
+        const parser = new Parser();
+        const ast = parser.astify(sql, { database: 'TransactSQL' });
```
Comment (jordanfanapour) [inline packages/SQLParser/src/cte-order-by-stripper.ts]: I don't think this should be hardcoded to "TransactSQL". It should be generalized to work on any sql dialect.
